aws/hll/aws-s3-transfer-manager/src/download/body.rs

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+use crate::download::worker::ChunkResponse;
+use crate::error::TransferError;
+use aws_smithy_types::byte_stream::AggregatedBytes;
+use std::cmp;
+use std::cmp::Ordering;
+use std::collections::BinaryHeap;
+use tokio::sync::mpsc;
+
+/// Stream of binary data representing an object's contents.
+///
+/// Wraps potentially multiple streams of binary data into a single coherent stream.
+/// The data on this stream is sequenced into the correct order.
+#[derive(Debug)]
+pub struct Body {
+    inner: UnorderedBody,
+    sequencer: Sequencer,
+}
+
+type BodyChannel = mpsc::Receiver<Result<ChunkResponse, TransferError>>;
+
+impl Body {
+    /// Create a new empty Body
+    pub fn empty() -> Self {
+        Self::new_from_channel(None)
+    }
+
+    pub(crate) fn new(chunks: BodyChannel) -> Self {
+        Self::new_from_channel(Some(chunks))
+    }
+
+    fn new_from_channel(chunks: Option<BodyChannel>) -> Self {
+        Self {
+            inner: UnorderedBody::new(chunks),
+            sequencer: Sequencer::new(),
+        }
+    }
+
+    /// Convert this body into an unordered stream of chunks.
+    pub(crate) fn unordered(self) -> UnorderedBody {
+        self.inner
+    }
+
+    /// Pull the next chunk of data off the stream.
+    ///
+    /// Returns [None] when there is no more data.
+    /// Chunks returned from a [Body] are guaranteed to be sequenced
+    /// in the right order.
+    pub async fn next(&mut self) -> Option<Result<AggregatedBytes, TransferError>> {
+        // TODO(aws-sdk-rust#1159, design) - do we want ChunkResponse (or similar) rather than AggregatedBytes? Would
+        //  make additional retries of an individual chunk/part more feasible (though theoretically already exhausted retries)
```

Comment:
Question: as "already exhausted retries" says, fetching individual chunks already benefits retries [as part of this call](https://github.com/smithy-lang/smithy-rs/pull/3697/files#diff-b91e2788fd16643b8df16c99701cd4cade84f33318fd08d063230fe1aba4400bR74-R78), right? (and if fetching one chunk fails after retries, we eventually error out at [a later point](https://github.com/smithy-lang/smithy-rs/pull/3697/files#diff-1bd049795ee294cf2e26a25903e7eef0ac42b813f0afd68652244f75bcaf2e57R140), in which case `unwrap` will panic)

Reply:
> we eventually error out at [a later point](https://github.com/smithy-lang/smithy-rs/pull/3697/files#diff-1bd049795ee294cf2e26a25903e7eef0ac42b813f0afd68652244f75bcaf2e57R140), in which case unwrap will panic

Right, this is just in the example driver program though.

I think this TODO comment was made thinking about pause/resume as well as if we are downloading a massive file and a few chunks fail including retries. Having the flexibility to potentially keep going vs what we have now where it would be very difficult to retry for just that missing chunk. We'll see how it plays out.